### ISSUE TITLE
[Feature]: Support Qwen3 Embedding & Reranker

### ISSUE THREAD

**Issue by ccdv-ai** [vllm-project/vllm#19229]
### 🚀 The feature, motivation and pitch

Core architecture is Qwen3 (0.6B, 4B, 8B).

The reranker relies on a specific logic to rank (softmax between logits of "yes" and "no" tokens):
https://huggingface.co/Qwen/Qwen3-Reranker-0.6B

The embedding model is more standard:
https://huggingface.co/Qwen/Qwen3-Embedding-0.6B

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by realsigridjin** [comment]
Look at this one for embedding usage.

```
# Requires vllm>=0.8.5
# https://github.com/QwenLM/Qwen3-Embedding?tab=readme-ov-file#vllm-usage
import logging
from typing import Dict, Optional, List

import json
import logging

import torch

from transformers import AutoTokenizer, is_torch_npu_available
from vllm import LLM, SamplingParams
from vllm.distributed.parallel_state import destroy_model_parallel
import gc
import math
from vllm.inputs.data import TokensPrompt

def format_instruction(instruction, query, doc):
    text = [
        {"role": "system", "content": "Judge whether the Document meets the requirements based on the Query and the Instruct provided. Note that the answer can only be \"yes\" or \"no\"."},
        {"role": "user", "content": f"<Instruct>: {instruction}\n\n<Query>: {query}\n\n<Document>: {doc}"}
    ]
    return text

def process_inputs(pairs, instruction, max_length, suffix_tokens):
    messages = [format_instruction(instruction, query, doc) for query, doc in pairs]
    messages =  tokenizer.apply_chat_template(
        messages, tokenize=True, add_generation_prompt=False, enable_thinking=False
    )
    messages = [ele[:max_length] + suffix_tokens for ele in messages]
    messages = [TokensPrompt(prompt_token_ids=ele) for ele in messages]
    return messages

def compute_logits(model, messages, sampling_params, true_token, false_token):
    outputs = model.generate(messages, sampling_params, use_tqdm=False)
    scores = []
    for i in range(len(outputs)):
        final_logits = outputs[i].outputs[0].logprobs[-1]
        token_count = len(outputs[i].outputs[0].token_ids)
        if true_token not in final_logits:
            true_logit = -10
        else:
            true_logit = final_logits[true_token].logprob
        if false_token not in final_logits:
            false_logit = -10
        else:
            false_logit = final_logits[false_token].logprob
        true_score = math.exp(true_logit)
        false_score = math.exp(false_logit)
        score = true_score / (true_score + false_score)
        scores.append(score)
    return scores

number_of_gpu = torch.cuda.device_count()
tokenizer = AutoTokenizer.from_pretrained('Qwen/Qwen3-Reranking-4B')
model = LLM(model='Qwen/Qwen3-Reranking-0.6B', tensor_parallel_size=number_of_gpu, max_model_len=10000, enable_prefix_caching=True, gpu_memory_utilization=0.8)
tokenizer.padding_side = "left"
tokenizer.pad_token = tokenizer.eos_token
suffix = "<|im_end|>\n<|im_start|>assistant\n<think>\n\n</think>\n\n"
max_length=8192
suffix_tokens = tokenizer.encode(suffix, add_special_tokens=False)
true_token = tokenizer("yes", add_special_tokens=False).input_ids[0]
false_token = tokenizer("no", add_special_tokens=False).input_ids[0]
sampling_params = SamplingParams(temperature=0, 
    max_tokens=1,
    logprobs=20, 
    allowed_token_ids=[true_token, false_token],
)

        
task = 'Given a web search query, retrieve relevant passages that answer the query'
queries = ["What is the capital of China?",
    "Explain gravity",
]
documents = [
    "The capital of China is Beijing.",
    "Gravity is a force that attracts two bodies towards each other. It gives weight to physical objects and is responsible for the movement of planets around the sun.",
]

pairs = list(zip(queries, documents))
inputs = process_inputs(pairs, task, max_length-len(suffix_tokens), suffix_tokens)
scores = compute_logits(model, inputs, sampling_params, true_token, false_token)
print('scores', scores)

destroy_model_parallel()
```

**Comment by czhcc** [comment]
vllm=0.8.5
Qwen3-Embedding可以通过/v1/embeddings调用
Qwen3-Reranker不行，用/v1/rerank调用返回The model does not support Rerank (Score) API

**Comment by noooop** [comment]
Qwen3-Embedding needs this to get the correct result.
https://huggingface.co/Qwen/Qwen3-Embedding-0.6B/discussions/2
https://github.com/embeddings-benchmark/mteb/pull/2769#issuecomment-2944905730

**Comment by lovetian1991** [comment]
> vllm=0.8.5 Qwen3-Embedding可以通过/v1/embeddings调用 Qwen3-Reranker不行，用/v1/rerank调用返回The model does not support Rerank (Score) API

这个会不会考虑兼容下 Rerank (Score) API

**Comment by sinopec** [comment]
embedding works in 0.9.0.1 while set task=embed, rerank failed because of  " 'Qwen3ForCausalLM' object has no attribute 'pooler'"

**Comment by noooop** [comment]
> embedding works in 0.9.0.1 while set task=embed, rerank failed because of " 'Qwen3ForCausalLM' object has no attribute 'pooler'"

working on it

**Comment by ccdv-ai** [comment]
The reranker model can be converted to a single label classification model:
https://huggingface.co/Qwen/Qwen3-Reranker-0.6B/discussions/3

**Comment by SeiunSky0131** [comment]
> vllm=0.8.5 Qwen3-Embedding可以通过/v1/embeddings调用 Qwen3-Reranker不行，用/v1/rerank调用返回The model does not support Rerank (Score) API

Could you give an example about how to call /v1/embeddings using curl command? I tried several different methods but found all of them failed. I do use vllm=0.8.5. Many thanks!

**Comment by 44670** [comment]
for qwen3 ranker models, a batched API that prefills the prompt and then returns the output logits of selected tokens (yes/no) would be helpful.

do we have such an api for now?

**Comment by xiakj** [comment]
mark

**Comment by vaclcer** [comment]
is the ranker function really fixed? I am still getting 

`'Qwen3ForCausalLM' object has no attribute 'pooler'`

error on v0.9.1

edit: 

ohh, I see... https://github.com/vllm-project/vllm/pull/19260#issuecomment-2968911059

**Comment by dengcao** [comment]
You can use Vllm to deploy the Qwen3-Embedding&Qwen3-Reranker large model.There is already a temporary solution:
以下是通过vllm快捷安装Qwen3-Embedding-0.6B/4B/8B，Qwen3-Reranker-0.6B/4B/8B的办法。

Docker 快速部署（一键安装包）：
Docker Compose Quick Installation:

Hugging Face：
https://huggingface.co/dengcao/Qwen3-Embedding-0.6B
https://huggingface.co/dengcao/Qwen3-Embedding-4B
https://huggingface.co/dengcao/Qwen3-Embedding-8B
https://huggingface.co/dengcao/Qwen3-Reranker-0.6B
https://huggingface.co/dengcao/Qwen3-Reranker-4B
https://huggingface.co/dengcao/Qwen3-Reranker-8B

ModelScope：
https://www.modelscope.cn/models/dengcao/Qwen3-Embedding-0.6B
https://www.modelscope.cn/models/dengcao/Qwen3-Embedding-4B
https://www.modelscope.cn/models/dengcao/Qwen3-Embedding-8B
https://www.modelscope.cn/models/dengcao/Qwen3-Reranker-0.6B
https://www.modelscope.cn/models/dengcao/Qwen3-Reranker-4B
https://www.modelscope.cn/models/dengcao/Qwen3-Reranker-8B

GitHub：
https://github.com/dengcao/Qwen3-Reranker-8B
https://github.com/dengcao/Qwen3-Reranker-4B
https://github.com/dengcao/Qwen3-Reranker-0.6B

**Comment by MaoJianwei** [comment]
> You can use Vllm to deploy the Qwen3-Embedding&Qwen3-Reranker large model.There is already a temporary solution: 以下是通过vllm快捷安装Qwen3-Embedding-0.6B/4B/8B，Qwen3-Reranker-0.6B/4B/8B的办法。
> 
> Docker 快速部署（一键安装包）： Docker Compose Quick Installation:
> 
> Hugging Face： https://huggingface.co/dengcao/Qwen3-Embedding-0.6B https://huggingface.co/dengcao/Qwen3-Embedding-4B https://huggingface.co/dengcao/Qwen3-Embedding-8B https://huggingface.co/dengcao/Qwen3-Reranker-0.6B https://huggingface.co/dengcao/Qwen3-Reranker-4B https://huggingface.co/dengcao/Qwen3-Reranker-8B
> 
> ModelScope： https://www.modelscope.cn/models/dengcao/Qwen3-Embedding-0.6B https://www.modelscope.cn/models/dengcao/Qwen3-Embedding-4B https://www.modelscope.cn/models/dengcao/Qwen3-Embedding-8B https://www.modelscope.cn/models/dengcao/Qwen3-Reranker-0.6B https://www.modelscope.cn/models/dengcao/Qwen3-Reranker-4B https://www.modelscope.cn/models/dengcao/Qwen3-Reranker-8B
> 
> GitHub： https://github.com/dengcao/Qwen3-Reranker-8B https://github.com/dengcao/Qwen3-Reranker-4B https://github.com/dengcao/Qwen3-Reranker-0.6B



做了哪些改动？让vllm可以支持Qwen3 Reranker模型

**Comment by forrestbao** [comment]
> 
> 做了哪些改动？让vllm可以支持Qwen3 Reranker模型

I think the magic here is setting the `task` to `score` when starting vllm.
